src/main/java/org/cyclonedx/gradle/utils/ExternalReferencesUtil.java

```diff
@@ -0,0 +1,85 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) OWASP Foundation. All Rights Reserved.
+ */
+package org.cyclonedx.gradle.utils;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.util.Objects;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.cyclonedx.model.Component;
+import org.cyclonedx.model.ExternalReference;
+
+public class ExternalReferencesUtil {
+
+    private ExternalReferencesUtil() {}
+
+    public static void complementByEnvironment(@Nonnull final Component component) {
+        // ignore all other VCSs for the time being
+        addGitReference(component);
+    }
+
+    private static void addGitReference(@Nonnull final Component component) {
+        // abort early if a VCS external reference has already been provided
+        if (component.getExternalReferences() != null
+                && component.getExternalReferences().stream()
+                        .filter(Objects::nonNull)
+                        .anyMatch(er -> ExternalReference.Type.VCS == er.getType())) {
+            return;
+        }
+
+        String gitUrl = fromEnvironment();
+
+        if (gitUrl == null || gitUrl.isEmpty()) {
+            gitUrl = fromGitRepo();
+        }
+
+        if (gitUrl != null) {
+            if (gitUrl.startsWith("git@")) {
+                gitUrl = "ssh://" + gitUrl;
+            }
+
+            final ExternalReference externalReference = new ExternalReference();
+            externalReference.setType(ExternalReference.Type.VCS);
+            externalReference.setUrl(gitUrl);
+
+            // addExternalReferences only accepts 'https://' and 'ssh://' remote urls.
+            // invalid urls will be ignored and the reference will not be added
+            component.addExternalReference(externalReference);
+        }
+    }
+
+    private static @Nullable String fromEnvironment() {
+        return System.getenv("GIT_URL");
+    }
+
+    private static @Nullable String fromGitRepo() {
+        try {
+            Process process = Runtime.getRuntime().exec("git remote get-url origin");
+
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+                return reader.readLine();
+            }
+        } catch (Exception ignore) {
+            // throw it away if this is not a git repository or if git is not installed
+        }
+
+        return null;
+    }
+}
```

Comment:
Minor suggestion - while `Runtime.getRuntime().exec(...` works well in most cases I have run into issues with not using a StreamGobbling technique:

```suggestion
    private static @Nullable String fromGitRepo() {
        try {
            ProcessBuilder processBuilder = new ProcessBuilder("git", "remote", "get-url", "origin");
            Process process = processBuilder.start();

            StreamGobbler outputGobbler = new StreamGobbler(process.getInputStream());
            StreamGobbler errorGobbler = new StreamGobbler(process.getErrorStream());

            outputGobbler.start();
            errorGobbler.start();

            int exitCode = process.waitFor();

            outputGobbler.join();
            errorGobbler.join();

            if (exitCode == 0) {
                return outputGobbler.getOutput().trim();
            }
        } catch (Exception ignore) {
            // throw it away if this is not a git repository or if git is not installed
        }
        return null;
    }

    private static class StreamGobbler extends Thread {
        private final InputStream inputStream;
        private final StringBuilder output = new StringBuilder();

        StreamGobbler(InputStream inputStream) {
            this.inputStream = inputStream;
        }

        @Override
        public void run() {
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    output.append(line);
                }
            } catch (IOException e) {
                // Handle or log exception as needed
            }
        }

        String getOutput() {
            return output.toString();
        }
    }
}
```

Reply:
Hey. I would like to know more about those issues you experienced. 

I'd rather go for simplicity if possible, and add no git reference if I run into a low probability issue. But if your issues occur often, the change may make sense.

Reply:
This is likely overkill for this shell command. However, in other cases you can run into issues if one of the streams (error or output) fills up it can cause the process to block/hang. 

This is a very [common technique](https://www.google.com/search?q=java+streamgobbler+process) used to read from a process.